### DISCUSSION TITLE
Mandatory parameter/header is missing: name

### DISCUSSION THREAD

**Discussion by evgeny-mordyasov** [line/armeria#5202] [Q&A] [answered]
I have two following methods:

```
    @Get
    @ProducesJson
    public HttpResponse findAllForBoard(@Param Long boardId) {
       ...
    }

    @Get
    @ProducesJson
    public HttpResponse findByName(@Param String name, @Param Long boardId) {
       ...
    }
```

Armeria:
```
    @Bean
    public ArmeriaServerConfigurator armeriaConfigurator() {
        return serverBuilder ->
                serverBuilder
                        .annotatedService("/tasks", restTaskService);
    }

```
When I make a request `localhost:8080/tasks?boardId=1`, I get the following error: `java.lang.IllegalArgumentException: Mandatory parameter/header is missing: name`.

When I make a request `localhost:8080/tasks?name=abc&boardId=1`,  there is no error.

How do I solve this problem?

**Log:**
WARN  com.linecorp.armeria.server.RejectedRouteHandler: Virtual host '*' has a duplicate route: /tasks

**Stacktrace:**
java.lang.IllegalArgumentException: Mandatory parameter/header is missing: name
com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.defaultOrException(AnnotatedValueResolver.java:1038)
com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.lambda$resolver$22(AnnotatedValueResolver.java:745)
com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.resolve(AnnotatedValueResolver.java:1013)
com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.lambda$toArguments$1(AnnotatedValueResolver.java:163)
java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.toArguments(AnnotatedValueResolver.java:163)
com.linecorp.armeria.internal.server.annotation.AnnotatedService.invoke(AnnotatedService.java:363)
com.linecorp.armeria.internal.server.annotation.AnnotatedService.lambda$serve1$3(AnnotatedService.java:327)
java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
com.linecorp.armeria.common.DefaultContextAwareRunnable.run(DefaultContextAwareRunnable.java:45)
java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
java.base/java.lang.Thread.run(Thread.java:833)

**Comment by jrhee17** [comment, marked as answer]
Hi! We probably need to add a `MatchesParam` to avoid the duplicate route.

I've confirmed the following works in my local env.
```java
@RegisterExtension
static final ServerExtension server = new ServerExtension() {
    @Override
    protected void configure(ServerBuilder sb) throws Exception {
        sb.annotatedService(new Object() {
            @Get("/")
            @MatchesParam("boardId")
            @ProducesJson
            public HttpResponse findAllForBoard(@Param Long boardId) {
                return HttpResponse.of(200);
            }

            @Get("/")
            @ProducesJson
            @MatchesParam("name")
            @MatchesParam("boardId")
            public HttpResponse findByName(@Param String name, @Param Long boardId) {
                return HttpResponse.of(201);
            }
        });
    }
};

@Test
void testDuplicate() {
    AggregatedHttpResponse res =
            server.blockingWebClient().get("/",
                                           QueryParams.of("boardId", "123"));
    assertThat(res.status().code()).isEqualTo(200);
    res = server.blockingWebClient().get("/",
                                         QueryParams.of("name", "jrhee17",
                                                        "boardId", "123"));
    assertThat(res.status().code()).isEqualTo(201);
}
```

I do think it might be more helpful if:
- We pick up `@Param` automatically when determining route duplication instead of having to explicitly specify `@MatchesParam`
- The default route duplication warning log compares strings, which may be misleading (we seem to be comparing `toString`)
  - https://github.com/line/armeria/blob/d40829d0d5c842ff18622319abd9f3c2eeab861e/core/src/main/java/com/linecorp/armeria/server/RejectedRouteHandler.java#L53

Let me bring this up with the team in the next business day

**Reply by evgeny-mordyasov** [reply]
Thank you. It really solved my problem. Please reflect this in the documentation for the "Annotated services" block.
